### PR TITLE
[ao][sparsity] L1 norm based block data sparsifier

### PR DESCRIPTION
**Pull request by macandro96**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79535
* __->__ #79534
* #79254
* #79253
* #79427
* #79252
* #79251

L1-Norm Sparsifier
This sparsifier computes the *L1-norm* of every sparse block and "zeroes-out" the
ones with the lowest norm. The level of sparsity defines how many of the
blocks is removed.
This sparsifier is controlled by three variables:
1. `sparsity_level` defines the number of *sparse blocks* that are zeroed-out
2. `sparse_block_shape` defines the shape of the sparse blocks. Note that
    the sparse blocks originate at the zero-index of the tensor.
3. `zeros_per_block` is the number of zeros that we are expecting in each
    sparse block. By default we assume that all elements within a block are
    zeroed-out. However, setting this variable sets the target number of
    zeros per block. The zeros within each block are chosen as the *smallest
    absolute values*.

Test Plan:
```python test/test_ao_sparsity.py TestNormDataSparsifiers```

Differential Revision: [D37164115](https://our.internmc.facebook.com/intern/diff/D37164115)

### CODE REVIEW

File: torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        mask = self.get_mask(name)
+        if sparsity_level <= 0 or zeros_per_block == 0:
+            mask.data = torch.ones_like(mask)
+        elif sparsity_level >= 1.0 and (zeros_per_block == values_per_block):
+            mask.data = torch.zeros_like(mask)
+
+        # Fetch the high level mask that zeros out entire blocks
+        data_lvl_mask = self._get_data_level_mask(data=data_norm, sparsity_level=sparsity_level,
+                                                  sparse_block_shape=sparse_block_shape)
+
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py:133]:
```
if values_per_block == zeros_per_block:
  We are done
```

File: torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        assert norm in ['L1'], "only L1 norm supported at the moment"
+
+        defaults = {'sparsity_level': sparsity_level, 'sparse_block_shape': sparse_block_shape,
+                    'zeros_per_block': zeros_per_block}
+        self.norm = norm
+        super().__init__(data_list=data_list, **defaults)
+
+    def _get_scatter_folded_mask(self, data, dim, indices, output_size, sparse_block_shape):
+        mask = torch.ones_like(data)
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py:47]: dtype should either bool or int or int8

File: torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        # Assume data is a squeezed tensor
+        height, width = data.shape[-2], data.shape[-1]
+        block_height, block_width = sparse_block_shape
+        # creating additional height and width to support padding
+        dh = (block_height - height % block_height) % block_height
+        dw = (block_width - width % block_width) % block_width
+
+        # create a new padded tensor like data (to match the block_shape)
+        padded_data = torch.ones(height + dh, width + dw, dtype=data.dtype, device=data.device)
+        padded_data = padded_data * float('inf')  # can also be replaced with 0 to stop the removal of edge data
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py:65]: try using `nan`

File: torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        unfolded_data = F.unfold(padded_data.unsqueeze(0).unsqueeze(0), kernel_size=sparse_block_shape,
+                                 stride=sparse_block_shape)
+
+        _, sorted_idx = torch.sort(unfolded_data, dim=1)
+        sorted_idx = sorted_idx[:, 0:zeros_per_block, :]  # zero out zeros_per_block number of elements
+
+        mask = self._get_scatter_folded_mask(data=unfolded_data, dim=1, indices=sorted_idx, output_size=padded_data.shape,
+                                             sparse_block_shape=sparse_block_shape)
+
+        mask = mask.squeeze()[0:height, 0:width]  # remove padding
```
Comment (z-a-f) [inline torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py:76]:
```
mask = mask.squeeze()[0:height, 0:width].contiguous()
```

**Review by z-a-f — APPROVED** [review]
Some comments

File: test/ao/sparsity/test_data_sparsifier.py [hunk 5]
```diff
@@ -230,3 +232,209 @@ def test_nn_embeddings(self):
...
         emb_test.run_tests()
+
+
+class TestNormDataSparsifierType(TestBaseDataSparsiferType):
+    def __init__(self, data_list, defaults, data_with_config, norm_type='L1'):
+        super().__init__(data_list=data_list, defaults=defaults, data_with_config=data_with_config)
+        assert norm_type in ['L1']
+        self.norm_type = norm_type
+
+    def get_bounds_on_actual_sparsity(self, config, tensor_shape):
```
Comment (HDCharles) [inline test/ao/sparsity/test_data_sparsifier.py:243]: if these are helper classes for other tests, should probably prepend `_` to signify it being private.

File: test/ao/sparsity/test_data_sparsifier.py [hunk 6]
```diff
@@ -230,3 +232,209 @@ def test_nn_embeddings(self):
         emb_test = TestBaseDataSparsiferType(data_list=data_list, defaults=defaults,
                                              data_with_config=data_with_config)
         emb_test.run_tests()
+
+
+class TestNormDataSparsifierType(TestBaseDataSparsiferType):
```
Comment (HDCharles) [inline test/ao/sparsity/test_data_sparsifier.py:237]: Similar to https://github.com/pytorch/pytorch/pull/79252/files#r897301980 I'm not understanding the name here, feels like it should be something like TestBaseDataSparsifier(Generator/Helper/Runner) or something like that and it should have a comment similar to the comment for the base class describing this.

File: torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py (hunk 1 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py:140]: done

File: torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        block_height, block_width = sparse_block_shape
+        # creating additional height and width to support padding
+        dh = (block_height - height % block_height) % block_height
+        dw = (block_width - width % block_width) % block_width
+
+        # create a new padded tensor like data (to match the block_shape)
+        padded_data = torch.ones(height + dh, width + dw, dtype=data.dtype, device=data.device)
+        padded_data = padded_data * float('inf')  # can also be replaced with 0 to stop the removal of edge data
+        padded_data[0:height, 0:width] = data
+        unfolded_data = F.unfold(padded_data.unsqueeze(0).unsqueeze(0), kernel_size=sparse_block_shape,
```
[the remark this answers is not in the document]
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py]: done

File: torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py (hunk 4 above)
Comment (macandro96) [reply torch/ao/sparsity/experimental/data_sparsifier/data_norm_sparsifier.py]: done

File: test/ao/sparsity/test_data_sparsifier.py (hunk 6 above)
Comment (macandro96) [reply test/ao/sparsity/test_data_sparsifier.py]: Renamed it to TestDataNormSparsifiersRunner and added comments

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164115).

File: test/ao/sparsity/test_data_sparsifier.py [hunk 8]
```diff
@@ -235,3 +237,216 @@ def test_nn_embeddings(self):
...
+            sparse_mask = (sparsified_data == 0).float()
+            if zpb == 0:
+                assert sparse_mask.mean() == 0
+            else:
+                # Ratio of individual zeros in the tensor
+                true_sl = min(max(sl, 0.0), 1.0)
+                true_sl = true_sl * zpb / sbs[0] / sbs[1]
+                assert sparse_mask.mean() == true_sl
+
+    def run_tests(self):
```
Comment (jerryzh168) [inline test/ao/sparsity/test_data_sparsifier.py:375]: I feel this is structured differently from what we have in other code, for example in https://github.com/pytorch/pytorch/blob/master/test/quantization/fx/test_quantize_fx.py#L702 we just define a TestCase and then some methods in the TestCase class which implements a single test. can we align with that structure?

Comment (macandro96) [reply test/ao/sparsity/test_data_sparsifier.py:375]: Do you mean that we convert each function inside the runner class into a separate class and invoke each one of them in the main test case class?

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164115).

File: test/ao/sparsity/test_data_sparsifier.py (hunk 8 above)
Comment (macandro96) [reply test/ao/sparsity/test_data_sparsifier.py:379]: added a TODO comment, will change structure in a later PR

**Comment by macandro96** [comment]
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164115).

**Pull request closed without merging**